commons/src/main/java/io/aiven/kafka/connect/common/source/AbstractSourceRecordIterator.java

```diff
@@ -0,0 +1,358 @@
+     *
+     * @param offset
+     *            the native key to start from. May be {@code null}.
+     * @return A stream of natvie objects. May be empty but not {@code null}.
+     */
+    abstract protected Stream<N> getNativeItemStream(K offset);
+
+    /**
+     * Gets an IOSupplier for the specific source record.
+     *
+     * @param sourceRecord
+     *            the source record to get the input stream from.
+     * @return the IOSupplier that retrieves an InputStream from the source record.
+     */
+    abstract protected IOSupplier<InputStream> getInputStream(T sourceRecord);
+
+    /**
+     * Gets the native key for the native object.
+     *
+     * @param nativeObject
+     *            the native object ot retrieve the native key for.
+     * @return The native key for the native object.
+     */
+    abstract protected K getNativeKey(N nativeObject);
+
+    /**
+     * Gets the AbstractSourceRecord implementation for the native object.
+     *
+     * @param nativeObject
+     *            the native object to get the AbstractSourceRecord for.
+     * @return the AbstractSourceRecord for the natvie object.
+     */
+    abstract protected T createSourceRecord(N nativeObject);
+
+    /**
+     * Creates an OffsetManagerEntry for a native object.
+     *
+     * @param nativeObject
+     *            the native object to create the OffsetManagerEntry for.
+     * @return An OffsetManagerEntry for a native object.
+     */
+    abstract protected O createOffsetManagerEntry(N nativeObject);
+
+    /**
+     * Creates an offset manager key for the native key.
+     *
+     * @param nativeKey
+     *            THe native key to create an offset manager key for.
+     * @return An offset manager key.
+     */
+    abstract protected OffsetManager.OffsetManagerKey getOffsetManagerKey(@NotNull K nativeKey);
+
+    @Override
+    final public boolean hasNext() {
+        if (!outer.hasNext() && lastSeenNativeKey != null) {
+            // update the buffer to contain this new objectKey
+            ringBuffer.enqueue(lastSeenNativeKey);
+            // Remove the last seen from the offsetmanager as the file has been completely processed.
+            offsetManager.removeEntry(getOffsetManagerKey(lastSeenNativeKey));
+        }
+        if (!inner.hasNext() && !outer.hasNext()) {
+            inner = getNativeItemStream(ringBuffer.getOldest()).map(fileMatching)
+                    .filter(taskAssignment)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .iterator();
+        }
+        while (!outer.hasNext() && inner.hasNext()) {
+            outer = convert(inner.next()).iterator();
+        }
+        return outer.hasNext();
+    }
+
+    @Override
+    final public T next() {
+        return outer.next();
+    }
+
+    @Override
+    final public void remove() {
+        throw new UnsupportedOperationException("This iterator is unmodifiable");
```

Comment:
We probably don't need to override the default behaviour, which is `throw new UnsupportedOperationException("remove");`